physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90

```diff
@@ -0,0 +1,7377 @@
+         input_nml_file = input_nml_file, fn_nml = fn_nml, version=3,       &
+         iostat = ios)
+    
+    ! -----------------------------------------------------------------------
+    ! write version number and namelist to log file
+    ! -----------------------------------------------------------------------
+    if (me == master) then
+       write (logunit, *) " ================================================================== "
+       write (logunit, *) "gfdl_cloud_microphysics_nml_v3"
+    endif
+
+    ! -----------------------------------------------------------------------
+    ! initialize microphysics variables
+    ! -----------------------------------------------------------------------
+
+    if (.not. tables_are_initialized) call qs_init
+
+    call setup_mp
+
+    ! -----------------------------------------------------------------------
+    ! define various heat capacities and latent heat coefficients at 0 deg K
+    ! -----------------------------------------------------------------------
+
+    call setup_mhc_lhc (hydrostatic)
+
+end subroutine gfdl_cloud_microphys_v3_mod_init
+
+! =======================================================================
+! GFDL cloud microphysics driver
+! =======================================================================
+
+subroutine gfdl_cloud_microphys_v3_mod_driver (qv, ql, qr, qi, qs, qg, qa, qnl, qni, pt, wa, &
+        ua, va, delz, delp, gsize, dtm, hs, water, rain, ice, snow, graupel, &
+        hydrostatic, is, ie, ks, ke, q_con, cappa, consv_te, adj_vmr, te, dte, &
+        prefluxw, prefluxr, prefluxi, prefluxs, prefluxg, last_step, do_inline_mp)
+
+    implicit none
+
+    ! -----------------------------------------------------------------------
+    ! input / output arguments
+    ! -----------------------------------------------------------------------
+
+    integer, intent (in) :: is, ie, ks, ke
+
+    logical, intent (in) :: hydrostatic, last_step, consv_te, do_inline_mp
+
+    real, intent (in) :: dtm
+
+    real, intent (in), dimension (is:ie) :: hs, gsize
+
+    real, intent (in), dimension (is:ie, ks:ke) :: qnl, qni
+
+    real, intent (inout), dimension (is:ie, ks:ke) :: delp, delz, pt, ua, va, wa, te
+    real, intent (inout), dimension (is:ie, ks:ke) :: qv, ql, qr, qi, qs, qg, qa
+    real, intent (inout), dimension (is:ie, ks:ke) :: prefluxw, prefluxr, prefluxi, prefluxs, prefluxg
+
+    real, intent (inout), dimension (is:, ks:) :: q_con, cappa
+
+    real, intent (inout), dimension (is:ie) :: water, rain, ice, snow, graupel
+
+    real, intent (out), dimension (is:ie, ks:ke) :: adj_vmr
+
+    real (kind = r8), intent (out), dimension (is:ie) :: dte
+
+    ! -----------------------------------------------------------------------
+    ! major cloud microphysics driver
+    ! -----------------------------------------------------------------------
+
+    call mpdrv (hydrostatic, ua, va, wa, delp, pt, qv, ql, qr, qi, qs, qg, qa, &
```

Comment:
This is another scheme with 3 layers of drivers. Not ideal, but I assume that there is a good reason to keep all of them?